tools/pybind11Tools.cmake

```diff
@@ -36,27 +36,78 @@ function(select_cxx_standard)
   endif()
 endfunction()
 
-# Internal: find the appropriate LTO flag for this compiler
-macro(_pybind11_find_lto_flag output_var prefer_thin_lto)
-  if(${prefer_thin_lto})
-    # Check for ThinLTO support (Clang)
-    check_cxx_compiler_flag("-flto=thin" HAS_THIN_LTO_FLAG)
-    set(${output_var} $<${HAS_THIN_LTO_FLAG}:-flto=thin>)
+# Internal: try compiling with the given c++ and linker flags; if it works, set success_out to 1 and
+# set cxx_out and linker_out in the *caller*'s parent scope and return from the *caller* (this is a
+# macro, not a function!)
+function(_pybind11_check_cxx_and_linker_flags cxxflags linkerflags cxx_out linker_out success_out)
+  set(CMAKE_REQUIRED_QUIET 1)
+  set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} ${linkerflags})
+  check_cxx_compiler_flag("${cxxflags}" has_lto)
+  if(has_lto)
+    separate_arguments(cxxflags)
+    separate_arguments(linkerflags)
+    set(${cxx_out} ${cxx_flags} PARENT_SCOPE)
+    set(${linker_out} ${linker_flags} PARENT_SCOPE)
+    set(${success_out} 1 PARENT_SCOPE)
+  else()
+    set(${success_out} 0 PARENT_SCOPE)
   endif()
+endfunction()
 
-  if(NOT ${prefer_thin_lto} OR NOT HAS_THIN_LTO_FLAG)
-    if(NOT CMAKE_CXX_COMPILER_ID MATCHES "Intel")
-      # Check for Link Time Optimization support (GCC/Clang)
-      check_cxx_compiler_flag("-flto" HAS_LTO_FLAG)
-      set(${output_var} $<${HAS_LTO_FLAG}:-flto>)
-    else()
-      # Intel equivalent to LTO is called IPO
-      check_cxx_compiler_flag("-ipo" HAS_IPO_FLAG)
-      set(${output_var} $<${HAS_IPO_FLAG}:-ipo>)
-    endif()
+# Wraps the above in a macro that resets the flags in the parent scope and returns if the check
+# succeeded (since this is a macro, not a function, the set/return is from the caller's POV)
+macro(_pybind11_return_if_cxx_and_linker_flags_work feature cxxflags linkerflags cxx_out linker_out)
+  _pybind11_check_cxx_and_linker_flags(${cxxflags} ${linkerflags} ${cxx_out} ${linker_out} check_success)
+  if(check_success)
+    set(${cxx_out} ${${cxx_out}} PARENT_SCOPE)
+    set(${linker_out} ${${linker_out}} PARENT_SCOPE)
+    message(STATUS "${feature} enabled (using ${cxxflags})")
+    return()
   endif()
 endmacro()
 
+# Internal: find the appropriate LTO flag for this compiler
+function(_pybind11_find_lto_flag cxxflags_out linkerflags_out prefer_thin_lto)
+  # Release type (upper case for simpler comparisons)
+  string(TOUPPER "${CMAKE_BUILD_TYPE}" U_CMAKE_BUILD_TYPE)
+
+  if(U_CMAKE_BUILD_TYPE MATCHES REL AND CMAKE_CXX_COMPILER_ID MATCHES "^GNU|Clang$") # *Not* ^(GNU|Clang)$
```

Comment:
It's irrelevant (and gone) in the updated version.  It now never checks it at all, and uses a generator expression to enable the LTO flags (except when doing a debug build).